sql: introduce crdb_internal.cluster_locks virtual table

AlexTalks opened the pull request:
This change introduces the new virtual table
`crdb_internal.cluster_locks`, as proposed in the KV Lock Observability
RFC (#75541), to expose lock contention.  This virtual table displays
the locks currently tracked in the lock tables in ranges across a
cluster, utilizing the KV `QueryLocksRequest` API to gather information
on the lock holders as well as the operations waiting on the locks
before converting to a user-friendly SQL view that incorporates
information about the database, table, schema, and index.

For example,
```
root@localhost:26261/defaultdb> select range_id, table_name, lock_key_pretty, txn_id, ts, lock_strength, durability, granted, contended, duration from crdb_internal.cluster_locks;
  range_id | table_name |     lock_key_pretty      |                txn_id                |             ts             | lock_strength | durability | granted | contended |    duration
-----------+------------+--------------------------+--------------------------------------+----------------------------+---------------+------------+---------+-----------+------------------
       235 | kv         | /Table/115/1/"alex"/0    | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.916745 | Exclusive     | Replicated |  true   |   true    | 00:01:50.607386
       235 | kv         | /Table/115/1/"alex"/0    | d71ea8eb-21b9-48a0-98b6-82eed84ed76f | 2022-03-26 00:24:02.137125 | None          | Replicated |  false  |   true    | 00:01:50.607376
       235 | kv         | /Table/115/1/"bob"/0     | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.916745 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607384
       235 | kv         | /Table/115/1/"chris"/0   | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.916745 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607383
       255 | kv         | /Table/115/1/"lauren"/0  | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   true    | 00:01:50.607757
       255 | kv         | /Table/115/1/"lauren"/0  | d71ea8eb-21b9-48a0-98b6-82eed84ed76f | 2022-03-26 00:24:02.137125 | None          | Replicated |  false  |   true    | 00:01:50.60773
       255 | kv         | /Table/115/1/"marilyn"/0 | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607755
       255 | kv         | /Table/115/1/"mike"/0    | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607754
       255 | kv         | /Table/115/1/"nancy"/0   | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607745
       255 | kv         | /Table/115/1/"noah"/0    | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607744
       255 | kv         | /Table/115/1/"paul"/0    | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607744
       255 | kv         | /Table/115/1/"peter"/0   | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607743
       256 | kv         | /Table/115/1/"sam"/0     | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.740235 | Exclusive     | Replicated |  true   |   true    | 00:01:50.607582
       256 | kv         | /Table/115/1/"sam"/0     | d71ea8eb-21b9-48a0-98b6-82eed84ed76f | 2022-03-26 00:24:02.137125 | None          | Replicated |  false  |   true    | 00:01:50.607573
       256 | kv         | /Table/115/1/"thomas"/0  | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.740235 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607581
       256 | kv         | /Table/115/1/"zebra"/0   | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.740235 | Exclusive     | Replicated |  true   |   false   | 00:01:50.60758
(16 rows)
```

This internal table is usable for system tenants as well as secondary
tenants.  It is also useful in conjunction with other
`crdb_internal.cluster_*` tables used for observability, and also can be
joined with itself to visualize blocked operations, and the transactions
that are blocking them.  This feature is gated on v22.1, to ensure that
older nodes will not receive `kv.Batch` requests with
`QueryLocksRequest`s in them.

Closes #74834

Release justification: Low risk, high benefit change.

Release note (sql change): introducing the `crdb_internal.cluster_locks`
virtual table, which exposes the current state of locks on keys tracked
by concurrency control. The virtual table displays metadata on locks
currently held by transactions, as well as operations waiting to obtain
the locks, and as such can be used to visualize active contention.
The `VIEWACTIVITY` or `VIEWACTIVITYREDACTED` role option, or the
`admin` role, is required to access the virtual table; however, if the
user only has the `VIEWACTIVITYREDACTED` role option, the key on which
a lock is held will be redacted.

AlexTalks committed:
sql: introduce crdb_internal.cluster_locks virtual table

This change introduces the new virtual table
`crdb_internal.cluster_locks`, as proposed in the KV Lock Observability
RFC (#75541), to expose lock contention.  This virtual table displays
the locks currently tracked in the lock tables in ranges across a
cluster, utilizing the KV `QueryLocksRequest` API to gather information
on the lock holders as well as the operations waiting on the locks
before converting to a user-friendly SQL view that incorporates
information about the database, table, schema, and index.

For example,
```
root@localhost:26261/defaultdb> select range_id, table_name, lock_key_pretty, txn_id, ts, lock_strength, durability, granted, contended, duration from crdb_internal.cluster_locks;
  range_id | table_name |     lock_key_pretty      |                txn_id                |             ts             | lock_strength | durability | granted | contended |    duration
-----------+------------+--------------------------+--------------------------------------+----------------------------+---------------+------------+---------+-----------+------------------
       235 | kv         | /Table/115/1/"alex"/0    | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.916745 | Exclusive     | Replicated |  true   |   true    | 00:01:50.607386
       235 | kv         | /Table/115/1/"alex"/0    | d71ea8eb-21b9-48a0-98b6-82eed84ed76f | 2022-03-26 00:24:02.137125 | None          | Replicated |  false  |   true    | 00:01:50.607376
       235 | kv         | /Table/115/1/"bob"/0     | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.916745 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607384
       235 | kv         | /Table/115/1/"chris"/0   | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.916745 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607383
       255 | kv         | /Table/115/1/"lauren"/0  | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   true    | 00:01:50.607757
       255 | kv         | /Table/115/1/"lauren"/0  | d71ea8eb-21b9-48a0-98b6-82eed84ed76f | 2022-03-26 00:24:02.137125 | None          | Replicated |  false  |   true    | 00:01:50.60773
       255 | kv         | /Table/115/1/"marilyn"/0 | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607755
       255 | kv         | /Table/115/1/"mike"/0    | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607754
       255 | kv         | /Table/115/1/"nancy"/0   | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607745
       255 | kv         | /Table/115/1/"noah"/0    | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607744
       255 | kv         | /Table/115/1/"paul"/0    | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607744
       255 | kv         | /Table/115/1/"peter"/0   | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.819704 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607743
       256 | kv         | /Table/115/1/"sam"/0     | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.740235 | Exclusive     | Replicated |  true   |   true    | 00:01:50.607582
       256 | kv         | /Table/115/1/"sam"/0     | d71ea8eb-21b9-48a0-98b6-82eed84ed76f | 2022-03-26 00:24:02.137125 | None          | Replicated |  false  |   true    | 00:01:50.607573
       256 | kv         | /Table/115/1/"thomas"/0  | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.740235 | Exclusive     | Replicated |  true   |   false   | 00:01:50.607581
       256 | kv         | /Table/115/1/"zebra"/0   | 7471e67b-bbc3-4dd9-a42f-43ec97949b14 | 2022-03-26 00:23:45.740235 | Exclusive     | Replicated |  true   |   false   | 00:01:50.60758
(16 rows)
```

This internal table is usable for system tenants as well as secondary
tenants.  It is also useful in conjunction with other
`crdb_internal.cluster_*` tables used for observability, and also can be
joined with itself to visualize blocked operations, and the transactions
that are blocking them.  This feature is gated on v22.1, to ensure that
older nodes will not receive `kv.Batch` requests with
`QueryLocksRequest`s in them.

Closes #74834

Release justification: Low risk, high benefit change.

Release note (sql change): introducing the `crdb_internal.cluster_locks`
virtual table, which exposes the current state of locks on keys tracked
by concurrency control. The virtual table displays metadata on locks
currently held by transactions, as well as operations waiting to obtain
the locks, and as such can be used to visualize active contention.
The `VIEWACTIVITY` or `VIEWACTIVITYREDACTED` role option, or the
`admin` role, is required to access the virtual table; however, if the
user only has the `VIEWACTIVITYREDACTED` role option, the key on which
a lock is held will be redacted.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77876)
<!-- Reviewable:end -->

kevin-v-ngo commented:
FYI - @xinhaoz

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/crdb_internal.go, line 5751 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MyL0JBY9nmeXHcf7ZOx:-MyL0JBY9nmeXHcf7ZOy:b-7m6dhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac516076f760d0825233166e6fd444373f5eab17/pkg/sql/crdb_internal.go#L5751)):*
> ```Go
> 
> 			// TODO(sarkesian): set key limits correctly, currently 5 used for testing.
> 			b.Header.MaxSpanRequestKeys = 5
> ```

@yuzefovich FYI - this is the logic I was talking about.  If it's possible to incorporate a limit hint into a request to a virtual table, definitely let me know!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 5751 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MyL0JBY9nmeXHcf7ZOx:-MyNbNweDniCquvtT2Qd:bkyfx7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac516076f760d0825233166e6fd444373f5eab17/pkg/sql/crdb_internal.go#L5751)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

@yuzefovich FYI - this is the logic I was talking about.  If it's possible to incorporate a limit hint into a request to a virtual table, definitely let me know!
</blockquote></details>

I see. Unfortunately, without a lot of plumbing we probably cannot push any limits / hints from the optimizer into this scope, so I'd just use the defaults we use elsewhere (I think 100k keys limit and 10MiB bytes limit).


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 5751 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MyL0JBY9nmeXHcf7ZOx:-Myo92bWCGz49hOa0lLG:bho8b9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac516076f760d0825233166e6fd444373f5eab17/pkg/sql/crdb_internal.go#L5751)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I see. Unfortunately, without a lot of plumbing we probably cannot push any limits / hints from the optimizer into this scope, so I'd just use the defaults we use elsewhere (I think 100k keys limit and 10MiB bytes limit).
</blockquote></details>

OK - got it.  I've just used the static constant values here.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
as discussed during team meeting, there's one `crdb_` too many in the name.

But the one thing that worries me a bit more is that the code currently does not paginate: all the results are accumulated in RAM before  the generator function is returned.

Usually, a vtable generator is so written that the underlying results are pulled incrementally as the rows get generated. This ensures we don't get more data in RAM than is consumed from the vtable, _and_ also ensures the memory accounting for the result rows is more-or-less aligned with that of the data source.
Here, the current implementation model breaks that. so at the very least the code should do its own memory accounting. And I'd encourage you to see how to make this code more "streaming-like".

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-:-MyoVIK7D2FrLOfiRS63:b-yzlwqx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

ajwerner commented:
This thing would really benefit from https://github.com/cockroachdb/cockroach/issues/74806. I'd go so far as to say we should try to get that issue scheduled if only to benefit features like this.

AlexTalks reviewed (commented):
@knz I've updated this a bit more so that a) it is more clear how the pagination works, and only one page of locks returned from `QueryLocksRequest` should be kept in memory at one time, and b) we query the correct spans, rather than simply `TableDataMin-TableDataMax` as before (which wouldn't support tenants).

Unless I'm unclear about how the generator works, I believe this should only perform a request when necessary - let me know if this is not correct.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-:-Myugey46pp6qQNRlzLT:b8y4k4t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Yes you have successfully addressed my concern. I'd welcome some additional explanatory comments (also for our future selves) though.

> Reviewed 1 of 2 files at r1, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-:-MywoUXs0ChPxpIv7RO6:bfim9pw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[-- commits, line 10 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MywnnZoB5fATe0Jeyx1):*
nit: be sure to expand this commit message a bit before this merges.
___
*[pkg/sql/crdb_internal.go, line 3110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mywo4x11ni7zgA3Zaz8:-Mywo4x11ni7zgA3Zaz9:b-mmk78n) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2ca1afef87c7e3e630b39a53d081277d6eea87/pkg/sql/crdb_internal.go#L3110)):*
> ```Go
> 	map[uint32]uint32,
> ) {
> 	// TODO(knz): maybe this could use internalLookupCtx.
> ```

Can you talk with folk on the SQL schema team to provide suggestions.
___
*[pkg/sql/crdb_internal.go, line 5688 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MywnesRAo5tt2qND9RV:-MywnesRAo5tt2qND9RW:b-7f77ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2ca1afef87c7e3e630b39a53d081277d6eea87/pkg/sql/crdb_internal.go#L5688)):*
> ```Go
> 		expensive operation since it creates a cluster-wide RPC-fanout.`,
> 	schema: `
> CREATE TABLE crdb_internal.crdb_locks (
> ```

are you still sure about this name?
___
*[pkg/sql/crdb_internal.go, line 5740 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MywoOcN6_YP00aDwZ9e:-MywoOcN6_YP00aDwZ9f:b-vz9dwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2ca1afef87c7e3e630b39a53d081277d6eea87/pkg/sql/crdb_internal.go#L5740)):*
> ```Go
> 		var resumeSpan *roachpb.Span
> 
> 		fetchLocks := func(key, endKey roachpb.Key) error {
> ```

I would welcome a few explanatory comments through this function to explain what is going on and how the code ensures the results are paginated.


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @knz)

___
*[pkg/sql/crdb_internal.go, line 3110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mywo4x11ni7zgA3Zaz8:-MyyO3AU2UPB6kDUZ50Q:bhy9ost) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2ca1afef87c7e3e630b39a53d081277d6eea87/pkg/sql/crdb_internal.go#L3110)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you talk with folk on the SQL schema team to provide suggestions.
</blockquote></details>

Will do!
___
*[pkg/sql/crdb_internal.go, line 5688 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MywnesRAo5tt2qND9RV:-MyyOVqMD9LfFsGFFGyn:b-m87ehx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2ca1afef87c7e3e630b39a53d081277d6eea87/pkg/sql/crdb_internal.go#L5688)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

are you still sure about this name?
</blockquote></details>

Very strange, I thought I had pushed this with my updated commit that changes the name to `cluster_locks` and also supports the secondary tenant use case.  Will double check.
___
*[pkg/sql/crdb_internal.go, line 5740 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MywoOcN6_YP00aDwZ9e:-MyyOjSoDolpKklVAw4t:b-9pu6y1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2ca1afef87c7e3e630b39a53d081277d6eea87/pkg/sql/crdb_internal.go#L5740)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would welcome a few explanatory comments through this function to explain what is going on and how the code ensures the results are paginated.
</blockquote></details>

I believe I've added some in my updated commit, hopefully that will help explain a bit.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
I just want to point out that it seems that we want to get this into 22.1 release, but this PR bumps the cockroach version. Once we cut the first 22.1.0-beta, I think bumps are no longer allowed, so we either have to slowdown the release train or this feature will have to wait for 22.2. The latter outcome seems rather unfortunate, so it'd be good to communicate with the release team, probably opening an issue with `release-blocker` (not `GA-blocker`) label would be good.

RichardJCai commented:
> I just want to point out that it seems that we want to get this into 22.1 release, but this PR bumps the cockroach version. Once we cut the first 22.1.0-beta, I think bumps are no longer allowed, so we either have to slowdown the release train or this feature will have to wait for 22.2. The latter outcome seems rather unfortunate, so it'd be good to communicate with the release team, probably opening an issue with `release-blocker` (not `GA-blocker`) label would be good.

Alternatively we could probably get away with not having the new version / version gate here. It seems like we're adding the version gate out of caution since it is an internal table.

AlexTalks commented:
> > I just want to point out that it seems that we want to get this into 22.1 release, but this PR bumps the cockroach version. Once we cut the first 22.1.0-beta, I think bumps are no longer allowed, so we either have to slowdown the release train or this feature will have to wait for 22.2. The latter outcome seems rather unfortunate, so it'd be good to communicate with the release team, probably opening an issue with `release-blocker` (not `GA-blocker`) label would be good.
> 
> Alternatively we could probably get away with not having the new version / version gate here. It seems like we're adding the version gate out of caution since it is an internal table.

The version gate is primarily to avoid sending `QueryLocksRequest` RPCs to nodes that don't support it (<22.1).  If this is not necessary, that's definitely fine and I can remove it!

ajwerner commented:
>The version gate is primarily to avoid sending QueryLocksRequest RPCs to nodes that don't support it (<22.1). If this is not necessary, that's definitely fine and I can remove it!

It sounds necessary to me.

yuzefovich commented:
The version gate sounds necessary to me too.

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @knz, and @yuzefovich)

___
*[pkg/clusterversion/cockroach_versions.go, line 550 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2dUL-4OoEOGVSQ7yH:-Mz2dUL-4OoEOGVSQ7yI:b7pfjpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/60e0f8e6d379ff3f13b09f7141df15948272c648/pkg/clusterversion/cockroach_versions.go#L550)):*
> ```Go
> 	},
> 	{
> 		Key:     ClusterLocksVirtualTable,
> ```

@rafiss @richardjcai please let me know if this bit is correct or should be removed!


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> Reviewed 4 of 4 files at r5, 4 of 14 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @knz, and @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 3108 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2nran1VgMBuXGVp40:-Mz2nraoCO7YsWB7dOq0:bvqnivz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L3108)):*
> ```Go
> 	map[uint32]map[uint32]string,
> 	map[uint32]uint32,
> 	map[uint32]uint32,
> ```

small nit: do you mind name the return variables of this function and the one below? 
___
*[pkg/sql/crdb_internal.go, line 5718 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2pvEj1ZVMWmFTiHaQ:-Mz2pvEj1ZVMWmFTiHaR:b-b9sqtr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5718)):*
> ```Go
> 			return nil, nil, err
> 		}
> 		if !hasViewActivityOrViewActivityRedacted {
> ```

When user have `ViewActivityRedacted` permission, we should ensure that they cannot view any PII information. That means we probably should omit `lock_key` and `lock_key_pretty` column.
___
*[pkg/sql/crdb_internal.go, line 5783 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2rMqn8j29v40bogo4:-Mz2rMqn8j29v40bogo5:bglb1lu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5783)):*
> ```Go
> 
> 			err := p.txn.Run(ctx, &b)
> 			if err == nil {
> ```

I feel we usually invert this conditions, so the error handling is closer to the call site  e.g.
 
``` golang 
if err != nil {
  return

resp = b.RawResponse().Responses[0].GetQueryLocks()
locks = resp.Locks
resumeSpan = resp.ResumeSpan
```
___
*[pkg/sql/crdb_internal.go, line 5813 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2siA1-55hfSLrLBDC:-Mz2siA1-55hfSLrLBDD:buajvz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5813)):*
> ```Go
> 					lockIdx = 0
> 				}
> 			}
> ```

preference nit: perhaps refactor the for loop into something like `maybeFetchLockInfo()` :thinking: 
___
*[pkg/sql/crdb_internal.go, line 5832 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2tPXeD5ZlXbPVQqNC:-Mz2tPXeD5ZlXbPVQqND:b47b7ab) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5832)):*
> ```Go
> 		// each waiter, prior to moving onto the next lock (or fetching additional
> 		// results as necessary).
> 		return func() (tree.Datums, error) {
> ```

I think the patterns  wrap the returning `virtualTableGenerator` using `setupGenerator()`. This would allow the generator to run in a separate goroutine and allows for graceful query cancellation. 
___
*[pkg/sql/crdb_internal.go, line 5838 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2uUes0hhBCsPBpTb-:-Mz2uUes0hhBCsPBpTb0:b-3dunmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5838)):*
> ```Go
> 			}
> 
> 			// If we couldn't get any more locks from getNextLock(), we are finished
> ```

nit: s/are finished/have finished/ 


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @knz, @rafiss, @RichardJCai, and @yuzefovich)

___
*[pkg/clusterversion/cockroach_versions.go, line 550 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2dUL-4OoEOGVSQ7yH:-MzB3E6o6fKCeqJSYd3A:bim97vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/60e0f8e6d379ff3f13b09f7141df15948272c648/pkg/clusterversion/cockroach_versions.go#L550)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

@rafiss @richardjcai please let me know if this bit is correct or should be removed!
</blockquote></details>

this part lgtm!


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
It's cool to see this! Have you thought about how you'd like to test it? The behavior should be stable enough to exercise in a logic test.

> Reviewed 1 of 4 files at r5, 14 of 14 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-:-MzITbeX71kQDqf8hOM6:bk6s8kq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @knz, and @RichardJCai)

___
*[-- commits, line 60 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MzG0DFC6CU42jyLWyrv):*
Reminder to fill in this release note.
___
*[pkg/sql/crdb_internal.go, line 3108 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2nran1VgMBuXGVp40:-MzG0UZw-ve3TBvVNj_X:b-bqdsfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L3108)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

small nit: do you mind name the return variables of this function and the one below? 
</blockquote></details>

+1, and could you add a comment to this function describing what it does?
___
*[pkg/sql/crdb_internal.go, line 5704 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MzITjm-3G-UK7cWWQAx:-MzITjm-3G-UK7cWWQAy:b-jt49mo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5704)):*
> ```Go
> );`,
> 	indexes: nil,
> 	generator: func(ctx context.Context, p *planner, db catalog.DatabaseDescriptor, stopper *stop.Stopper) (virtualTableGenerator, cleanupFunc, error) {
> ```

Do you think it would be worthwhile to enumerate the set of `virtualIndex`s we plan to define on this table?

Also, consider writing this to minimize the amount of code that will need to move when we do add `virtualIndex`s. The flow of 1) get descriptors, 2) get spans, 3) get locks for spans is good, but we may want to extract parts of this out into reusable helpers.
___
*[pkg/sql/crdb_internal.go, line 5779 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MzITDN83BtNQZsvyoEu:-MzITDN83BtNQZsvyoEv:b-2xic6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5779)):*
> ```Go
> 			b.AddRawRequest(queryLocksRequest)
> 
> 			b.Header.MaxSpanRequestKeys = int64(rowinfra.ProductionKVBatchSize)
> ```

Should we be using a memory monitor here like we do in `crdbInternalClusterStmtStatsTable`?


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
I'm working on adding some more tests now, unfortunately it's difficult to encapsulate the concept of "waiters" in a logic test because the logic tests expect all statements to complete I believe...

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-:-MzMo2T57pOeYjI64A4O:b-89kfy4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @knz, and @RichardJCai)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
> because the logic tests expect all statements to complete I believe...

I think so too (but could be wrong). It sounds logic test framework is not the right one for this, probably you want to add regular unit tests.

AlexTalks commented:
@yuzefovich I may have a solution with #79010 - please take a look if you can.  This change has been rebased on that, so that more testing can be added.

AlexTalks committed:
kvserver: ensure correct durability prioritization in query lockholders

While previously when evaluating a `QueryLocksRequest` in the lock
table, the default was to look at the unreplicated lock holder first, we
should instead prioritize selecting the replicated lock holder as it is
of stronger durability.

Release note: None

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @knz, @nvanbenschoten, @rafiss, and @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 3108 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2nran1VgMBuXGVp40:-N-6MKSk-1iNmJHTQ3hc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L3108)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1, and could you add a comment to this function describing what it does?
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 5704 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MzITjm-3G-UK7cWWQAx:-N-6NsIi75oMjrqoEveo:be8lfu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5704)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do you think it would be worthwhile to enumerate the set of `virtualIndex`s we plan to define on this table?

Also, consider writing this to minimize the amount of code that will need to move when we do add `virtualIndex`s. The flow of 1) get descriptors, 2) get spans, 3) get locks for spans is good, but we may want to extract parts of this out into reusable helpers.
</blockquote></details>

Should be ready to check out #79623 !
___
*[pkg/sql/crdb_internal.go, line 5718 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2pvEj1ZVMWmFTiHaQ:-N-6Niq24xTPQC5Vtwxd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5718)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

When user have `ViewActivityRedacted` permission, we should ensure that they cannot view any PII information. That means we probably should omit `lock_key` and `lock_key_pretty` column.
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 5779 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MzITDN83BtNQZsvyoEu:-N-6O66-3w23hkXKfy9T:bo4ylxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5779)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we be using a memory monitor here like we do in `crdbInternalClusterStmtStatsTable`?
</blockquote></details>

I actually don't know the answer to this - do we have any guideline here?
___
*[pkg/sql/crdb_internal.go, line 5783 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2rMqn8j29v40bogo4:-N-6OUoV8XMRGZ4BTCcx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5783)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

I feel we usually invert this conditions, so the error handling is closer to the call site  e.g.
 
``` golang 
if err != nil {
  return

resp = b.RawResponse().Responses[0].GetQueryLocks()
locks = resp.Locks
resumeSpan = resp.ResumeSpan
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> Dismissed @Azhng from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @knz, @nvanbenschoten, @rafiss, and @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 3110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mywo4x11ni7zgA3Zaz8:-N08QiIg-qJmqTmHQc4T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2ca1afef87c7e3e630b39a53d081277d6eea87/pkg/sql/crdb_internal.go#L3110)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Will do!
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 5688 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MywnesRAo5tt2qND9RV:-N08QucmAe70r5u-eCTh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2ca1afef87c7e3e630b39a53d081277d6eea87/pkg/sql/crdb_internal.go#L5688)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Very strange, I thought I had pushed this with my updated commit that changes the name to `cluster_locks` and also supports the secondary tenant use case.  Will double check.
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 5740 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MywoOcN6_YP00aDwZ9e:-N08QwRBBz8_uO434c3U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2ca1afef87c7e3e630b39a53d081277d6eea87/pkg/sql/crdb_internal.go#L5740)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

I believe I've added some in my updated commit, hopefully that will help explain a bit.
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 5779 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MzITDN83BtNQZsvyoEu:-N-j8UQ8C1SuGH3rH0cy:b-l3q9yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5779)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

I actually don't know the answer to this - do we have any guideline here?
</blockquote></details>

Looking into this a bit more, I think it wouldn't hurt to have but I think we'll need to refactor this to create a worker and use `setupGenerator`.  If possible, I'd like to do this in a follow-up PR.
___
*[pkg/sql/crdb_internal.go, line 5832 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2tPXeD5ZlXbPVQqNC:-N08REwE-KnCcRWftRwh:b-37sk6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5832)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

I think the patterns  wrap the returning `virtualTableGenerator` using `setupGenerator()`. This would allow the generator to run in a separate goroutine and allows for graceful query cancellation. 
</blockquote></details>

OK.  Planning to do this in a follow-up PR.


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @knz, @nvanbenschoten, @rafiss, and @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 5779 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-MzITDN83BtNQZsvyoEu:-N0CxSVh9pzqksdaHMLA:b-z3vthz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5779)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Looking into this a bit more, I think it wouldn't hurt to have but I think we'll need to refactor this to create a worker and use `setupGenerator`.  If possible, I'd like to do this in a follow-up PR.
</blockquote></details>

Incorporated into #80216.
___
*[pkg/sql/crdb_internal.go, line 5832 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-Mz2tPXeD5ZlXbPVQqNC:-N0CxVZW4bkx1vUqpuHt:b-z3vthz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5ba1d0b724e2e00e9231843da21304b88423a4e/pkg/sql/crdb_internal.go#L5832)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

OK.  Planning to do this in a follow-up PR.
</blockquote></details>

Incorporated into #80216.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice job on this @alextalks. Let's get someone from SQL to sign off as well, then we can go ahead and merge.

> Reviewed 1 of 18 files at r10, 10 of 12 files at r14, 3 of 3 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-:-N0EFQMyDAcLE8RpdOxD:b719bju)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @knz, @nvanbenschoten, @rafiss, and @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 6019 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0EEoCb5r2rZbKqexiA:-N0EEoCc-pp2zrwhKXgH:b-68jqbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L6019)):*
> ```Go
> 			waiterIdx++
> 
> 			strippedKey, _, _ := keys.DecodeTenantPrefix(curLock.Key)
> ```

nit: do we even want to run this line and the next when `shouldRedactKeys`?
___
*[pkg/ccl/logictestccl/testdata/logic_test/cluster_locks_tenant, line 77 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0EF87v7_-qcWrl3eV2:-N0EF87v7_-qcWrl3eV3:b-ca4nv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/ccl/logictestccl/testdata/logic_test/cluster_locks_tenant#L77)):*
> ```
> 
> # looking at each transaction separately, validate the expected results in the lock table
> query TTTTTTBB colnames,retry
> ```

It's cool that we can write tests like these!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 1 of 18 files at r10, 11 of 11 files at r13, 10 of 12 files at r14, 3 of 3 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @knz, @nvanbenschoten, @rafiss, and @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 5840 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0Gt1S_1CDObe3P7mV0:-N0Gt1S_1CDObe3P7mV1:bkaubmv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L5840)):*
> ```Go
> 	indexes: nil,
> 	generator: func(ctx context.Context, p *planner, db catalog.DatabaseDescriptor, stopper *stop.Stopper) (virtualTableGenerator, cleanupFunc, error) {
> 		if !p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.ClusterLocksVirtualTable) {
> ```

nit: add a TODO to remove this gate in 22.2 time frame.
___
*[pkg/sql/crdb_internal.go, line 5929 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0GtgaO8Z-gBAAQy2AN:-N0GtgaO8Z-gBAAQy2AO:bnj7tzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L5929)):*
> ```Go
> 			}
> 
> 			resp = b.RawResponse().Responses[0].GetQueryLocks()
> ```

nit: should we be defensive and check the length of `b.RawResponse()` returning an error if it's not equal to 1?
___
*[pkg/sql/crdb_internal.go, line 5983 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0GuIdq7gkbiUOWpjDD:-N0GuIdq7gkbiUOWpjDE:b-u1dhl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L5983)):*
> ```Go
> 			// If we couldn't get any more locks from getNextLock(), we have finished
> 			// generating result rows.
> 			if curLock == nil {
> ```

nit: it seems cleaner to explicitly add `|| fErr != nil` check to the conditional.
___
*[pkg/sql/crdb_internal.go, line 5995 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0GucFg8TYeo3aifz0H:-N0GucFg8TYeo3aifz0I:b7m80z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L5995)):*
> ```Go
> 			if waiterIdx < 0 {
> 				if curLock.LockHolder != nil {
> 					txnIDDatum = tree.NewDUuid(tree.DUuid{UUID: curLock.LockHolder.ID})
> ```

nit: might be good to use `DatumAlloc` for the new datums. I guess we don't do this for other internal virtual tables, so feel free to ignore this comment.
___
*[pkg/sql/logictest/testdata/logic_test/cluster_locks, line 1 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0GxaLmAPS1BxyXwM7d:-N0GxaLmAPS1BxyXwM7e:buas3j1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/logictest/testdata/logic_test/cluster_locks#L1)):*
> ```
> # LogicTest: local 5node
> ```

I'm curious about the choice of these modes. IIUC this file is separate from `cluster_locks_tenant` because we manually split ranges. However, we then don't move those ranges to other nodes (if we're in `5node` config). My guess is that you want to verify the behavior in a distributed setting, and I don't think this file accomplishes that.


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @AlexTalks, @Azhng, @knz, @nvanbenschoten, @rafiss, and @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 5840 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0Gt1S_1CDObe3P7mV0:-N0HjWib0mjbX_ML4pkd:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L5840)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: add a TODO to remove this gate in 22.2 time frame.
</blockquote></details>

OK
___
*[pkg/sql/crdb_internal.go, line 5929 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0GtgaO8Z-gBAAQy2AN:-N0HliNE8KgoJe8SoGTy:bjwalw3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L5929)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we be defensive and check the length of `b.RawResponse()` returning an error if it's not equal to 1?
</blockquote></details>

OK - can add an assertion, though strangely I notice there's some other places in crdb_internal.go where we don't do this ¯\_(ツ)_/¯ 
___
*[pkg/sql/crdb_internal.go, line 5983 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0GuIdq7gkbiUOWpjDD:-N0HmAUp2acUR0iy_tIX:b-2pgfyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L5983)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it seems cleaner to explicitly add `|| fErr != nil` check to the conditional.
</blockquote></details>

This way `fErr` simply falls through, though I can add it as it doesn't hurt at all.  Had always learned/thought that Go semantics sort of requires you to return `(nil, non-nil err)` when your return params are `(*value, error)`.
___
*[pkg/sql/crdb_internal.go, line 6019 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0EEoCb5r2rZbKqexiA:-N0HnfcR2Bb_u3OEJLQQ:b-mh1sc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L6019)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: do we even want to run this line and the next when `shouldRedactKeys`?
</blockquote></details>

OK - I suppose it's not necessary at all so if it seems cleaner that's cool.
___
*[pkg/sql/logictest/testdata/logic_test/cluster_locks, line 1 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0GxaLmAPS1BxyXwM7d:-N0HoE_c1PfXBomMkMPk:b-quuhsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/logictest/testdata/logic_test/cluster_locks#L1)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I'm curious about the choice of these modes. IIUC this file is separate from `cluster_locks_tenant` because we manually split ranges. However, we then don't move those ranges to other nodes (if we're in `5node` config). My guess is that you want to verify the behavior in a distributed setting, and I don't think this file accomplishes that.
</blockquote></details>

Yes - I would have tried to put it in one file, but `skipif` only applies to statements and queries, which means I can't `skipif config 3node-tenant` the lines where I have `let $r1; select range_id...`.  Also, `local` and `5node` seemed to make sense because the test doesn't play well with the `fakedist` configs.

Your thought is that we should manually move these ranges around with `ALTER TABLE t EXPERIMENTAL_RELOCATE...`?  Likely in a `skipif config local` statement?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 1 of 1 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @AlexTalks, @Azhng, @knz, @nvanbenschoten, @rafiss, and @yuzefovich)

___
*[pkg/sql/crdb_internal.go, line 5983 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-N0GuIdq7gkbiUOWpjDD:-N0IJZ3136YSMDaiCMC2:b-yyn5rf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf0835985592b11dfe86a1ff1d500e06be0e4fc/pkg/sql/crdb_internal.go#L5983)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

This way `fErr` simply falls through, though I can add it as it doesn't hurt at all.  Had always learned/thought that Go semantics sort of requires you to return `(nil, non-nil err)` when your return params are `(*value, error)`.
</blockquote></details>

Hm, I guess it's a best practice to return `(nil, non-nil err)`, but I think I've also come across cases of returning `(x, non-nil err)` where `x` is undefined.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Note that there appear to be some more tests in need of [adjustment](https://teamcity.cockroachdb.com/viewLog.html?buildId=4972516&buildTypeId=Cockroach_UnitTests).

> Reviewed 2 of 2 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77876#-:-N0IKh6NBwA8crspkK2z:b-bigmio)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @AlexTalks, @Azhng, @knz, @nvanbenschoten, @rafiss, and @yuzefovich)



<!-- Sent from Reviewable.io -->

AlexTalks commented:
bors r+
